mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,319 @@
+        try:
+            output = subprocess.check_output(path_to_version_sh)
+        except OSError as e:
+            warn("Could not record emirge's package versions: " + str(e))
+
+    mgr.set_constant("emirge_package_versions", output)
+
+# }}}
+
+
+# {{{ State handling
+
+def set_sim_state(mgr: LogManager, dim, state, eos) -> None:
+    """Update the simulation state of all :class:`StateConsumer` of the log manager.
+
+    Parameters
+    ----------
+    mgr
+        The :class:`logpyle.LogManager` to set the state of.
+    """
+    state_vars = {}
+
+    for gd_lst in [mgr.before_gather_descriptors,
+            mgr.after_gather_descriptors]:
+        for gd in gd_lst:
+            if isinstance(gd.quantity, StateConsumer):
+                extract_state_vars_func = gd.quantity.extract_state_vars
+                if extract_state_vars_func not in state_vars:
+                    state_vars[extract_state_vars_func] = \
+                        extract_state_vars_func(dim, state, eos)
+
+                gd.quantity.set_state_vars(state_vars[extract_state_vars_func])
+
+
+class StateConsumer:
+    """Base class for quantities that require a state for logging.
+
+    .. automethod:: __init__
+    .. automethod:: set_state_vars
+    """
+
+    def __init__(self, extract_vars_for_logging: Callable):
+        """Store the function to extract state variables.
+
+        Parameters
+        ----------
+        extract_vars_for_logging
+            Returns a dict of the state vars for a particular state.
+        """
+        self.extract_state_vars = extract_vars_for_logging
+        self.state_vars = None
+
+    def set_state_vars(self, state_vars: np.ndarray) -> None:
+        """Update the state vector of the object."""
+        self.state_vars = state_vars
+
+# }}}
+
+# {{{ Discretization-based quantities
+
+
+class DiscretizationBasedQuantity(LogQuantity, StateConsumer):
+    """Logging support for physical quantities.
+
+    Possible rank aggregation operations (``op``) are: min, max, norm.
+    """
+
+    def __init__(self, discr: Discretization, quantity: str, op: str,
+                 extract_vars_for_logging, units_logging, name: str = None,
+                 axis: Optional[int] = None):
+        unit = units_logging(quantity)
+
+        if name is None:
+            name = f"{op}_{quantity}" + (str(axis) if axis is not None else "")
+
+        LogQuantity.__init__(self, name, unit)
+        StateConsumer.__init__(self, extract_vars_for_logging)
+
+        self.discr = discr
+
+        self.quantity = quantity
+        self.axis = axis
+
+        from functools import partial
+
+        if op == "min":
+            self._discr_reduction = partial(self.discr.nodal_min, "vol")
+            self.rank_aggr = min
+        elif op == "max":
+            self._discr_reduction = partial(self.discr.nodal_max, "vol")
+            self.rank_aggr = max
+        elif op == "norm":
+            self._discr_reduction = partial(self.discr.norm)
```

Comment:
```suggestion
            self._discr_reduction = partial(self.discr.norm, p=2)
```
(Otherwise, why is that `partial` there?)